gateway/src/lib.rs

```diff
@@ -237,29 +235,42 @@ pub trait EndStateExt<'c>: EndState<'c> {
     }
 }
 
-impl<'c, S> EndStateExt<'c> for S where S: EndState<'c> {}
+impl<Ctx, S> EndStateExt<Ctx> for S
+where
+    S: Clone + TryState + EndState<Ctx>,
+    Ctx: Send + Sync,
+{
+}
+
+/// A [`State`] which contains all its transitions, including
+/// failures
+pub trait TryState: Sized {
+    type ErrorVariant;
+
+    fn into_result(self) -> Result<Self, Self::ErrorVariant>;
+}
 
-pub trait IntoEndState<'c, E>
+pub trait IntoTryState<E>
 where
-    E: EndState<'c>,
+    E: TryState,
 {
-    fn into_end_state(self) -> Result<E, Infallible>;
+    fn into_try_state(self) -> Result<E, Infallible>;
```

Comment:
Maybe we should no longer call this generic `E` as it can be confused with e for error.